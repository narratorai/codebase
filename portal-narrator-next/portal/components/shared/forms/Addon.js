import React, { Component } from 'react'
import styled from 'styled-components'

export const StyledAddon = styled(({ content, disabled, left, ...props }) => (
  <div {...props}>
    <span>{content}</span>
  </div>
))`
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid ${(props) => props.theme.colors.gray300};
  border-right-width: ${(props) => (props.left ? '0px' : '1px')};
  border-left-width: ${(props) => (props.left ? '1px' : '0px')};
  background: ${(props) => (props.disabled ? props.theme.colors.gray100 : props.theme.colors.white)};

  &:hover,
  &:focus {
    cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
    border: 1px solid ${(props) => (props.disabled ? 'inherit' : props.theme.colors.blue300)};
    border-right-width: ${(props) => (props.left ? '0px' : '1px')};
    border-left-width: ${(props) => (props.left ? '1px' : '0px')};
  }

  > span {
    max-height: 16px;
  }

  svg,
  use {
    fill: ${(props) => (props.disabled ? props.theme.colors.gray300 : 'auto')};
  }
`

class Addon extends Component {
  handleClick = () => {
    const { disabled, onClick } = this.props

    if (disabled) {
      return true
    }

    if (onClick) {
      return onClick()
    }
  }

  render() {
    const { onClick, ...rest } = this.props

    return <StyledAddon onClick={this.handleClick} {...rest} />
  }
}

Addon.propTypes = {}

export default Addon
