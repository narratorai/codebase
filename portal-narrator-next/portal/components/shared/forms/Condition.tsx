import { includes, isEmpty } from 'lodash'
import React from 'react'
import { Field } from 'react-final-form'

interface ConditionProps {
  when: string
  is?: boolean
  isNot?: boolean
  isIn?: any[]
  exists?: boolean
  children: React.ReactNode
}

export const Condition = ({ when, is, isNot, isIn = [], children, exists }: ConditionProps) => (
  <Field name={when} subscription={{ value: true }}>
    {({ input: { value } }) => {
      // If you pass in props exists, you just care that the value has been entered already
      if (exists && !isEmpty(value)) {
        return children
      }

      // If you pass props.isIn, value could be isIn a collection
      if (!isEmpty(isIn) && includes(isIn, value)) {
        return children
      }

      // if you pass props.is, it's about equivalency
      if (value === is) {
        return children
      }

      // if you pass props.is, it's about equivalency
      if (isNot && value !== isNot) {
        return children
      }

      return null
    }}
  </Field>
)

export default Condition
