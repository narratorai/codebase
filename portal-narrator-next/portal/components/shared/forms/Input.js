import React, { Component } from 'react'
import _ from 'lodash'
import styled from 'styled-components'
import PropTypes from 'prop-types'

// DUPLICATE OF StyledTextarea
export const StyledInput = styled(({ disabled, meta, carryRef, ...props }) => (
  <input ref={carryRef} disabled={disabled} {...props} />
))`
  width: 100%;
  background: ${(props) => props.theme.colors.white};
  font-size: ${(props) => props.theme.fontSizes[2]}px;
  border: 1px solid
    ${(props) =>
      _.get(props, 'meta.error') && _.get(props, 'meta.submitFailed')
        ? props.theme.colors.red500
        : props.theme.colors.gray400};
  border-radius: 4px;
  line-height: 1.23;
  padding: 10px;

  &:hover,
  &:focus {
    &:not(:disabled) {
      cursor: pointer;
      border: 1px solid ${(props) => props.theme.colors.blue300};
      box-shadow: none;
    }
  }

  /* stylelint-disable-next-line no-descending-specificity */
  &:disabled {
    background: ${(props) => props.theme.colors.gray100};
    opacity: 1;

    &:hover {
      cursor: not-allowed;
    }
  }

  &::placeholder {
    color: ${(props) => props.theme.colors.gray500};
  }
`

class Input extends Component {
  // Fix issue where we would hit the enter key and it would submit the form
  // https://github.com/davidkpiano/react-redux-form/issues/790
  handleKeyPress = (event) => {
    const { allowEnterToSubmit } = this.props
    if (allowEnterToSubmit) {
      return
    }
    if (event.key === 'Enter') event.preventDefault()
  }

  render() {
    const { allowEnterToSubmit, carryRef, ...otherProps } = this.props
    return <StyledInput f={1} onKeyPress={this.handleKeyPress} carryRef={carryRef} {...otherProps} />
  }
}

Input.propTypes = {
  allowEnterToSubmit: PropTypes.bool.isRequired,
}
Input.defaultProps = {
  allowEnterToSubmit: false,
}

export default Input
