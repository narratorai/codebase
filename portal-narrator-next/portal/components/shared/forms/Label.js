import React from 'react'
import styled from 'styled-components'

const StyledLabel = styled(({ disabled, inline, large, strong, text, ...props }) => (
  <label disabled={disabled} {...props}>
    {text}
  </label>
))`
  display: inline-block;
  font-size: ${(props) => (props.large ? `${props.theme.fontSizes[1]}px` : `${props.theme.fontSizes[0]}px`)};
  color: ${(props) => (props.disabled ? props.theme.colors.gray500 : props.theme.colors.gray700)};
  line-height: 1.27;
  margin-bottom: ${(props) => (props.inline ? '0px' : '7px')};
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  font-weight: ${(props) => (props.strong ? 600 : 300)};
`

const Label = (props) => <StyledLabel {...props} />

export default Label
