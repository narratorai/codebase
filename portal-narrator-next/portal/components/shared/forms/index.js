export { default as Addon, StyledAddon } from './Addon'
export { default as Condition } from './Condition'
export { default as CustomerTableContextSelect } from './CustomerTableContextSelect'
export { default as Input, StyledInput } from './Input'
export { default as Label } from './Label'
export { default as SearchResult } from './Search/SearchResult'
export { default as Textarea } from './Textarea'
