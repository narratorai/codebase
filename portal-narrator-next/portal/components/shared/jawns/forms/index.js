export { default as BasicEditorWrapper } from './BasicEditorWrapper'
export { default as Condition } from './Condition'
export { default as CronTranslation } from './CronTranslation'
export { default as Input, inputStyle, RenameInput } from './Input'
export { default as Label } from './Label'
export { RenameTextarea, default as Textarea } from './Textarea'
