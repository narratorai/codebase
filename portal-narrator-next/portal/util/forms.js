import _ from 'lodash'
import { isEmail as validatorIsEmail, isURL } from 'validator'
import { isValidCron } from 'cron-validator'
import { notCompanySlugs } from '@/util/auth'
import { handleApproximateCron } from 'util/helpers'
// eslint-disable-next-line
import FinalFormArrayMutators from 'final-form-arrays'

///////////////// HELPERS /////////////////

/**
 * JSON.stringify with replacer that converts undefined values to null
 * @param arg - JSON.stringify argument
 */
export const stringifyWithUndefined = (arg) =>
  JSON.stringify(arg, (k, v) => {
    if (v === undefined) {
      return null
    }
    return v
  })

///////////////// VALIDATORS /////////////////

export const jsonValidation = (value) => {
  try {
    if (_.isString(value)) {
      JSON.parse(value)
    }
  } catch (error) {
    return 'Invalid JSON'
  }
  return value ? undefined : 'Required'
}

export const jsonArrayValidation = (value) => {
  const jsonErrors = jsonValidation(value)
  if (jsonErrors) {
    return jsonErrors
  }

  let arr = value
  if (_.isString(value)) {
    arr = JSON.parse(arr)
  }

  if (!Array.isArray(arr)) {
    return 'JSON must be an array'
  }
}

// Allow for positive/negative check as well as required
// Used by <NumberField />
export const numberValidator = ({ value, isPositiveOnly = false, isRequired = true }) => {
  if (isPositiveOnly) {
    const integerValue = _.toInteger(value)
    if (integerValue < 0) {
      return 'Cannot be negative'
    }
  }

  if (isRequired) {
    return required(value)
  }
}

export const cronValidator = ({ value, isRequired = true }) => {
  if (isRequired && required(value)) {
    return required(value)
  }

  // escape early if there is no value and it is not required
  if (!isRequired && _.isEmpty(value)) {
    return null
  }

  const { formattedCronTab } = handleApproximateCron(value)

  if (formattedCronTab && !isValidCron(formattedCronTab, { seconds: true })) {
    return 'Invalid Cron Tab'
  }

  return null
}

export const required = (value) => {
  // for arrays specifically
  if (_.isArray(value) && _.isEmpty(value)) {
    return 'Required'
  }
  // for numbers as 0 is considered false
  if (_.isNumber(value)) {
    return value || value === 0 ? undefined : 'Required'
  }
  return value ? undefined : 'Required'
}

export const isEmail = (value) => {
  const requiredError = required(value)
  if (requiredError) {
    return requiredError
  }
  return validatorIsEmail(value) ? undefined : 'Invalid Email'
}

export const areEmails = (emails, isRequired = false) => {
  const requiredError = required(emails)
  if (isRequired && requiredError) {
    return requiredError
  }

  const invalidEmails = []
  _.forEach(emails, (email) => {
    if (!validatorIsEmail(email)) {
      invalidEmails.push(email)
    }
  })

  return _.isEmpty(invalidEmails) ? undefined : `The following emails are invalid: ${_.join(invalidEmails, ', ')}`
}

export const isValidWebhookUrl = (value, message) => {
  const requiredError = required(value)
  if (requiredError) {
    return requiredError
  }

  return isURL(value, { protocols: ['https'], require_protocol: true })
    ? undefined
    : message || 'Webhook url requires HTTPS'
}

export const isValidKlaviyoListUrl = (value) => {
  const requiredError = required(value)
  if (requiredError) {
    return requiredError
  }

  return isURL(value, {
    protocols: ['https'],
    host_whitelist: ['www.klaviyo.com', 'mc.sendgrid.com'],
    require_protocol: true,
    disallow_auth: true,
  })
    ? undefined
    : 'Not a valid Klaviyo/Sendgrid URL.'
}

export const isUnique = (value, list) => {
  if (!value) return undefined

  return _.includes(list, value) ? `${value} already exists.` : undefined
}

export const isValidCompanySlug = (value) => {
  const slugRegex = /^[a-z][a-z0-9-]*[a-z0-9]$/
  return slugRegex.test(value)
    ? undefined
    : 'A slug can only contain lowercase letters, numbers, and dashes (-), must start with a letter, and cannot end with a dash'
}

export const isCompanySlug = (value) => {
  return composeValidators(required, isNotForbiddenCompanySlug, isValidCompanySlug)(value)
}

export const isNotForbiddenCompanySlug = (value) => {
  const lowerValue = _.toLower(value)
  if (_.includes(notCompanySlugs, lowerValue)) {
    return `${value} is a reserved name`
  }
}

// composes multiple validators into a single validation function
export const composeValidators =
  (...validators) =>
  (value) =>
    validators.reduce((error, validator) => error || validator(value), undefined)

// used to validate field configs in Build Narrative inputs
export const fieldValidator = (error) => (error?.message ? () => error.message : () => {})

// insert is very broken in narratives (and I imagine elsewhere)
// this overrides react-final-form-array's extra increment logic added to insert
// that breaks our forms
export const arrayMutators = {
  ...FinalFormArrayMutators,
  // https://github.com/final-form/react-final-form-arrays/issues/138#issuecomment-681958884
  insert: ([name, index, value], state, { changeValue }) => {
    changeValue(state, name, (array) => {
      const copy = [...(array || [])]
      copy.splice(index, 0, value)
      return copy
    })
  },
}
