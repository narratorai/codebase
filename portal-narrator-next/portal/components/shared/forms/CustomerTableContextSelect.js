import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import _ from 'lodash'

import { Label } from 'components/shared/forms'
import { Box, Flex } from 'components/shared/jawns'
import { useCompany } from 'components/context/company/hooks'

const CustomerTableContextSelect = ({ fieldName }) => {
  const company = useCompany()

  // Ignore choosing context if there's only v_customers
  if (_.isEmpty(company.customer_tables) || company.customer_tables.length === 1) {
    return null
  }

  return (
    <Flex>
      {_.map(company.customer_tables, (table) => (
        <Box key={table.name} mr="20px">
          <Flex alignItems="center">
            <Box mr="5px">
              <Field id={table.name} name={fieldName} type="radio" component="input" value={table.name} />
            </Box>
            <Label htmlFor={table.name} text={table.label} inline />
          </Flex>
        </Box>
      ))}
    </Flex>
  )
}

CustomerTableContextSelect.propTypes = {
  fieldName: PropTypes.string.isRequired,
}

export default CustomerTableContextSelect
