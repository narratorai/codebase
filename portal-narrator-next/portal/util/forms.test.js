import { stringifyWithUndefined } from './forms'

describe('stringifyWithUndefined()', () => {
  it('converts undefined values to null', () => {
    expect(
      stringifyWithUndefined({
        keep: 'this',
        this_too: undefined,
      })
    ).toMatchSnapshot()
  })
})
