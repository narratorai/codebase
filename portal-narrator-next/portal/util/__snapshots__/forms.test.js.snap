// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyWithUndefined() converts undefined values to null 1`] = `"{"keep":"this","this_too":null}"`;
